src/pip/_internal/req/constructors.py

```diff
@@ -57,6 +58,27 @@ def convert_extras(extras: Optional[str]) -> Set[str]:
     return get_requirement("placeholder" + extras.lower()).extras
 
 
+def _set_requirement_extras(req: Requirement, new_extras: Set[str]) -> Requirement:
+    """
+    Returns a new requirement based on the given one, with the supplied extras. If the
+    given requirement already has extras those are replaced (or dropped if no new extras
+    are given).
+    """
+    match: Optional[re.Match[str]] = re.fullmatch(
+        # see https://peps.python.org/pep-0508/#complete-grammar
+        r"([\w\t .-]+)(\[[^\]]*\])?(.*)",
+        str(req),
+        flags=re.ASCII,
+    )
+    # ireq.req is a valid requirement so the regex should always match
+    assert match is not None
+    pre: Optional[str] = match.group(1)
+    post: Optional[str] = match.group(3)
+    assert pre is not None and post is not None
```

Comment:
```suggestion
    pre: str
    post: str
    pre, post = match.group(1, 3)
```

Interestingly, Mypy (typeshed, technically) does not mark the return value of `group` as Optional, despite the documentation clears says None is a possible return value, so this would just work. I’m not totally sure if we should rely on this though.

Reply:
I found some context on [the motivation for the Any union](https://github.com/python/typeshed/issues/9465). I see why they couldn't make the change, but I still feel like clients of the library should ideally make an effort to use the correct type, even where typeshed is too permissive by necessity. Let me know if you think differently.